soroban-spec/src/gen/typescript.rs

```diff
@@ -0,0 +1,245 @@
+    VerifySha256 { expected: String },
+    #[error("parsing contract spec: {0}")]
+    Parse(stellar_xdr::Error),
+    #[error("getting contract spec: {0}")]
+    GetSpec(FromWasmError),
+}
+
+pub fn generate_from_file(
+    file: &str,
+    verify_sha256: Option<&str>,
+) -> Result<String, GenerateFromFileError> {
+    // Read file.
+    let wasm = fs::read(file).map_err(GenerateFromFileError::Io)?;
+
+    // Produce hash for file.
+    let sha256 = Sha256::digest(&wasm);
+    let sha256 = format!("{:x}", sha256);
+
+    if let Some(verify_sha256) = verify_sha256 {
+        if verify_sha256 != sha256 {
+            return Err(GenerateFromFileError::VerifySha256 { expected: sha256 });
+        }
+    }
+
+    // Generate code.
+    let code = generate_from_wasm(&wasm).map_err(GenerateFromFileError::GetSpec)?;
+    Ok(code)
+}
+
+pub fn generate_from_wasm(wasm: &[u8]) -> Result<String, FromWasmError> {
+    let spec = from_wasm(wasm)?;
+    let code = generate(&spec);
+    Ok(code)
+}
+
+pub fn generate(spec: &[ScSpecEntry]) -> String {
+    let header = r#"import * as SorobanClient from "soroban-client";
+let xdr = SorobanClient.xdr;
+
+// TODO: Move all the non-trivial conversions to js-stellar-base and use them them from there.
+function bigintToI128(value: bigint): SorobanClient.xdr.ScVal {
+  const buf = bigintToBuf(value);
+  if (buf.length > 16) {
+    throw new Error("value overflows i128");
+  }
+
+  if (value < BigInt(0)) {
+    // Clear the top bit
+    buf[0] &= 0x7f;
+  }
+
+  // left-pad with zeros up to 16 bytes
+  let padded = Buffer.alloc(16);
+  buf.copy(padded, padded.length-buf.length);
+
+  if (value < BigInt(0)) {
+    // Set the top bit
+    padded[0] |= 0x80;
+  }
+
+  const hi = new xdr.Uint64(
+    Number(bigintFromBytes(false, ...padded.slice(4, 8))),
+    Number(bigintFromBytes(false, ...padded.slice(0, 4)))
+  );
+  const lo = new xdr.Uint64(
+    Number(bigintFromBytes(false, ...padded.slice(12, 16))),
+    Number(bigintFromBytes(false, ...padded.slice(8, 12)))
+  );
+
+  return xdr.ScVal.scvObject(xdr.ScObject.scoI128(new xdr.Int128Parts({lo, hi})));
+}
+
+function bigintToU128(value: bigint): SorobanClient.xdr.ScVal {
+  const buf = bigintToBuf(value);
+  if (buf.length > 16) {
+    throw new Error("value overflows i128");
+  }
+
+  // left-pad with zeros up to 16 bytes
+  let padded = Buffer.alloc(16);
+  buf.copy(padded, padded.length-buf.length);
+
+  const hi = new xdr.Uint64(
+    Number(bigintFromBytes(false, ...padded.slice(4, 8))),
+    Number(bigintFromBytes(false, ...padded.slice(0, 4)))
+  );
+  const lo = new xdr.Uint64(
+    Number(bigintFromBytes(false, ...padded.slice(12, 16))),
+    Number(bigintFromBytes(false, ...padded.slice(8, 12)))
+  );
+
+  return xdr.ScVal.scvObject(xdr.ScObject.scoU128(new xdr.Int128Parts({lo, hi})));
+}
+
+function bigintFromBytes(signed: boolean, ...bytes: (string | number | bigint)[]): bigint {
+    let sign = 1;
+    if (signed && bytes[0] === 0x80) {
```

Comment:
❗ If I'm understanding this correctly this is checking the top byte, not just the top bit. I think you need to `&` with `0x80`.